src/vmm/src/device_manager/acpi.rs

```diff
@@ -0,0 +1,123 @@
+// Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use acpi_tables::{aml, Aml};
+use kvm_ioctls::VmFd;
+use vm_memory::GuestAddress;
+
+use super::resources::ResourceAllocator;
+use crate::devices::acpi::vmgenid::{VmGenId, VmGenIdError};
+use crate::vstate::memory::GuestMemoryMmap;
+
+#[derive(Debug, thiserror::Error, displaydoc::Display)]
+pub enum ACPIDeviceManagerError {
+    /// Failed to allocate requested resource: {0}
+    Allocator(#[from] vm_allocator::Error),
+    /// Error in VMGenID device: {0}
+    VmGenID(#[from] VmGenIdError),
+    /// Error registering interrupt: {0}
+    InterruptRegister(#[from] kvm_ioctls::Error),
+}
+
+#[derive(Debug)]
+pub struct ACPIDeviceManager {
+    /// VMGenID device
+    pub vmgenid: Option<VmGenId>,
+}
+
+impl ACPIDeviceManager {
+    /// Create a new ACPIDeviceManager object
+    pub fn new() -> Self {
+        Self { vmgenid: None }
+    }
+
+    /// Create a new VMGenID device
+    ///
+    /// This will allocate resources (guest memory and one interrupt line) for the device,
+    /// build the device and registers its interrupt line with KVM.
+    pub fn attach_vmgenid(
+        &mut self,
+        mem: &GuestMemoryMmap,
+        resource_allocator: &mut ResourceAllocator,
+        vm_fd: &VmFd,
+    ) -> Result<(), ACPIDeviceManagerError> {
+        let gsi = resource_allocator.allocate_gsi(1)?;
+        let addr = resource_allocator.allocate_system_memory(
+            4096,
+            8,
+            vm_allocator::AllocPolicy::LastMatch,
+        )?;
+        self.register_vmgenid(gsi[0], GuestAddress(addr), mem, vm_fd)
+    }
+
+    /// Register a new VMGenID device
+    ///
+    /// Create and register a VMGenID device using the provided guest address and GSI.
+    /// Register the GSI with KVM
+    ///
+    /// This will create the device and register the interrupt line with KVM.
+    pub fn register_vmgenid(
+        &mut self,
+        gsi: u32,
+        guest_address: GuestAddress,
+        mem: &GuestMemoryMmap,
+        vm_fd: &VmFd,
+    ) -> Result<(), ACPIDeviceManagerError> {
+        let vmgenid = VmGenId::new(guest_address, gsi, mem)?;
```

Comment:
We don't do that for other devices. With MMIO it's the device manager that allocates resources for the particular device and then does the registration. What if we rethink this logic (if there is a strong reason for that) when we do the device manager refactoring

Reply:
For other devices we do smth similar: we create a device first and then we put them in the device manager. The only reason I made `VmGenId` to allocate it's resources because there is no `MMIODeviceInfo` alternative for `acpi` devices. The `VmGenId` contains this info in itself, so I think it make sense for it to know what it needs.

Reply:
We do have the equivalent of `MMIODeviceInfo` for ACPI devices, it's the memory address and IRQ. For MMIO devices, we do allocate the memory and IRQ in the equivalent device manager. IMO, VMGenID needs just to know that it has an interrupt and an address, it doesn't need to know how to allocate them.

Reply:
In this case maybe there should be a `ACPIDeviceInfo` struct that has the resources and `VmGenId` can just contain this struct?
```rust
let acpi_info = vmm.acpi_device_manager.alloc_resource(&mut vmm.resource_allocator);
let vmgenid = VmGenId::new(&vmm.guest_memory, acpi_info);
vmm.acpi_device_manager.attach_vmgenid(&vmm.vm, vmgenid);

// in restore case the `acpi_info` can be constructed from pieces.
let acpi_info = AcpiInfo::new(irq, mem_region);
```
Just to rephrase my overall concern here: the way devices work is that they are created outside of device managers and only are inserted into device managers when it is needed. I don't mind device manager to allocate resources(irqs, mem regions) when device is being inserted into the device manager, but I don't like when device manager actually created a device and then adds it into itself in one function call. If all acpi devices will have same resources requirements (same amout if irqs and memory region size) then maybe smth link `AcpiInfo::new(&mut vmm.resource_allocator)` will work nicely. This way the interface for `ACPIDeviceManager` can be more generic like `ACPIDeviceManager::add_device<D: AcpiDevice>(device: D, acpi_info: AcpiInfo)`. 
Anyway I think I am getting into the depth of how devices should work. Let me open an issue so we can gather all our thoughts there.
For now I think what you have here does it's job.

Reply:
I don't see much value in creating a struct to store two variables inside. In MMIO this makes sense because later we store that struct somewhere (`MMIODeviceManager::id_to_dev_info`).

But I think you convinced me to do actually perform the allocation inside `Vmgenid`, because not all ACPI devices will need to perform the same allocations. I will do it like that.

Reply:
Don't want to hijack the conversation but my 2 cents on the device info:
`ACPIDeviceManager->VmGenId->guest_address` and `ACPIDeviceManager->VmGenId->gsi` though provide device info, they are still specific to vmgenid and not the acpi device manager itself which in the future can have more devices.
Some context on the next problem: vmgenid will also be used with acpi device manager but with devicetree and not with acpi support. To create devicetree nodes we will need vmgenid info so we need to pass the vmgenid addr/gsi form builder.rs all the way to fdt.rs.
Problem:
1. With current approach if we pass vmgenid to fdt.rs we pass more than just addr/gsi which fdt.rs doesn't need to know.
2. If we have more devices in the future which follows same approach of using ACPI device manager for devicetree, the hashmap `pub(crate) id_to_dev_info: HashMap<(DeviceType, String), MMIODeviceInfo>,` in MMIODeviceManager looks more reasonable for passing info to fdt than passing individual device info.
The 2nd issue, though revolves around hypothetical devices, shouldn't be hard to avoid with an ACPIDeviceInfo.